Introduce `O(n)` canonicalization algorithm

evanlinjin opened the pull request:
Fixes bitcoindevkit/bdk#1665
Replaces bitcoindevkit/bdk#1659

### Description

Previously, getting the canonical history of transactions/UTXOs required calling `TxGraph::get_chain_position` on each transaction. This was highly inefficient and resulted in an `O(n^2)` algorithm. The situation is especially problematic when we have many unconfirmed conflicts.

This PR introduces an `O(n)` algorithm to determine the canonical set of transactions in `TxGraph`. The algorithm's premise is as follows:

1. If transaction `A` is determined to be canonical, all of `A`'s ancestors must also be canonical.
2. If transaction `B` is determined to be NOT canonical, all of `B`'s descendants must also be NOT canonical.
3. If a transaction is anchored in the best chain, it is canonical.
4. If a transaction conflicts with a canonical transaction, it is NOT canonical.
5. A transaction with a higher last-seen has precedence.
6. Last-seen values are transitive. A transaction's collective last-seen value is the max of it's last-seen value and all of it's descendants.

We maintain two mutually-exclusive `txid` sets: `canoncial` and `not_canonical`.

Imagine a method `mark_canonical(A)` that is based on premise 1 and 2. This method will mark transaction `A` and all of it's ancestors as canonical. For each transaction that is marked canonical, we can iterate all of it's conflicts and mark those as `non_canonical`. If a transaction already exists in `canoncial` or `not_canonical`, we can break early, avoiding duplicate work.

This algorithm iterates transactions in 3 runs.

1. Iterate over all transactions with anchors in descending anchor-height order. For any transaction that has an anchor pointing to the best chain, we call `mark_canonical` on it. We iterate in descending-height order to reduce the number of anchors we need to check against the `ChainOracle` (premise 1). The purpose of this run is to populate `non_canonical` with all transactions that directly conflict with anchored transactions and populate `canonical` with all anchored transactions and ancestors of anchors transactions (transitive anchors).
2. Iterate over all transactions with last-seen values, in descending last-seen order. We can call `mark_canonical` on all of these that do not already exist in `canonical` or `not_canonical`.
3. Iterate over remaining transactions that contains anchors (but not in the best chain) and have no last-seen value. We treat these transactions in the same way as we do in run 2.

#### Benchmarks

Thank you to @ValuedMammal for working on this.

```sh
$ cargo bench -p bdk_chain --bench canonicalization
```

Benchmark results (this PR):

```
many_conflicting_unconfirmed::list_canonical_txs                                                                            
                        time:   [709.46 us 710.36 us 711.35 us]
many_conflicting_unconfirmed::filter_chain_txouts                                                                            
                        time:   [712.59 us 713.23 us 713.90 us]
many_conflicting_unconfirmed::filter_chain_unspents                                                                            
                        time:   [709.95 us 711.16 us 712.45 us]
many_chained_unconfirmed::list_canonical_txs                                                                             
                        time:   [2.2604 ms 2.2641 ms 2.2680 ms]
many_chained_unconfirmed::filter_chain_txouts                                                                             
                        time:   [3.5763 ms 3.5869 ms 3.5979 ms]
many_chained_unconfirmed::filter_chain_unspents                                                                             
                        time:   [3.5540 ms 3.5596 ms 3.5652 ms]
nested_conflicts_unconfirmed::list_canonical_txs                                                                            
                        time:   [660.06 us 661.75 us 663.60 us]
nested_conflicts_unconfirmed::filter_chain_txouts                                                                            
                        time:   [650.15 us 651.36 us 652.71 us]
nested_conflicts_unconfirmed::filter_chain_unspents                                                                            
                        time:   [658.37 us 661.54 us 664.81 us]
```

Benchmark results (master): https://github.com/evanlinjin/bdk/tree/fix/1665-master-bench

```
many_conflicting_unconfirmed::list_canonical_txs                                                                             
                        time:   [94.618 ms 94.966 ms 95.338 ms]
many_conflicting_unconfirmed::filter_chain_txouts                                                                             
                        time:   [159.31 ms 159.76 ms 160.22 ms]
many_conflicting_unconfirmed::filter_chain_unspents                                                                             
                        time:   [163.29 ms 163.61 ms 163.96 ms]

# I gave up running the rest of the benchmarks since they were taking too long.
```

### Notes to the reviewers

* ***PLEASE MERGE bitcoindevkit/bdk#1733 BEFORE THIS PR!*** We had to change the signature of `ChainPosition` to account for transitive anchors and unconfirmed transactions with no `last-seen` value.

* The canonicalization algorithm is contained in `/crates/chain/src/canonical_iter.rs`.

* Since the algorithm requires traversing transactions ordered by anchor height, and then last-seen values, we introduce two index fields in `TxGraph`; `txs_by_anchor` and `txs_by_last_seen`. Methods `insert_anchor` and `insert_seen_at` are changed to populate these index fields.

* An ADR is added: `docs/adr/0003_canonicalization_algorithm.md`. This is based on the work in bitcoindevkit/bdk#1592.

### Changelog notice

* Added: Introduce an `O(n)` canonicalization algorithm. This logic is contained in `/crates/chain/src/canonical_iter.rs`.
* Added: Indexing fields in `TxGraph`; `txs_by_anchor_height` and `txs_by_last_seen`. Pre-indexing allows us to construct the canonical history more efficiently.
* Removed: `TxGraph` methods: `try_get_chain_position` and `get_chain_position`. This is superseded by the new canonicalization algorithm.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
feat(chain)!: `TxGraph` contain anchors in one field

Previously, the field `TxGraph::anchors` existed because we assumed
there was use in having a partially-chronological order of transactions
there. Turns out it wasn't used at all.

This commit removes anchors from the `txs` field and reworks `anchors`
field to be a map of `txid -> set<anchors>`.

This is a breaking change since the signature of `all_anchors()` is
changed.

Update `TxGraph` field docs for `empty_outspends` and `empty_anchors`.

evanlinjin committed:
feat(chain)!: `O(n)` canonicalization algorithm

This is an O(n) algorithm to determine the canonical set of txids.

* Run 1: Iterate txs with anchors, starting from highest anchor height
  txs.
* Run 2: Iterate txs with last-seen values, starting from highest
  last-seen values.
* Run 3: Iterate txs that are remaining from run 1 which are not
  anchored in the best chain.

Since all transitively-anchored txs are added to the `canonical` set in
run 1, and anything that conflicts to anchored txs are already added to
`not_canonial`, we can guarantee that run 2 will not traverse anything
that directly or indirectly conflicts anything that is anchored.

Run 3 is needed in case a tx does not have a last-seen value, but is
seen in a conflicting chain.

`TxGraph` is updated to include indexes `txids_by_anchor_height` and
`txids_by_last_seen`. These are populated by the `insert_anchor` and
`insert_seen_at` methods. Generic constaints needed to be tightened as
these methods need to be aware of the anchor height to create
`LastSeenIn`.

notmandatory commented on crates/chain/src/tx_graph.rs:
```diff
@@ -135,15 +139,15 @@ impl<A: Ord + Clone> From<TxUpdate<A>> for TxGraph<A> {
+    anchors: HashMap<Txid, BTreeSet<A>>,
     last_seen: HashMap<Txid, u64>,
 
     // This atrocity exists so that `TxGraph::outspends()` can return a reference.
```
No need to call this an "atrocity" since it's the best way to do it.
```suggestion
    // This exists so that `TxGraph::outspends()` can return a reference to an empty set of `Txid`s.
```

evanlinjin replied:
~This is out of scope for this PR. I'll turn this into a ticket.~

I realized that this is not out of scope, since we need to explain the purpose of `empty_anchors` below and might as well do this together.

notmandatory commented on crates/chain/src/tx_graph.rs:
```diff
@@ -135,15 +139,15 @@ impl<A: Ord + Clone> From<TxUpdate<A>> for TxGraph<A> {
     // This atrocity exists so that `TxGraph::outspends()` can return a reference.
     // FIXME: This can be removed once `HashSet::new` is a const fn.
     empty_outspends: HashSet<Txid>,
+    empty_anchors: BTreeSet<A>,
```
```suggestion
    // This exists so that `TxGraph` functions can return a reference to an empty set of `TxNode.anchors`.
    empty_anchors: BTreeSet<A>,
```

notmandatory commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1050,10 +1050,17 @@ impl Wallet {
     ///         last_seen,
     ///     ),
+    ///     ChainPosition::UnconfirmedAndNotSeen => println!(
+    ///         "tx does not conflict with anything canonical, but we never seen it in mempool",
```
⛏️ grammar nit
```suggestion
    ///         "tx does not conflict with anything canonical, but we have never seen it in the mempool",
```

notmandatory commented:
Per call today @evanlinjin will open a new PR for the 1.0.0-beta milestone that only makes expected breaking changes for `Wallet`,  likely only the `ChainPosition` enum. Then the algorithm improvements in this PR can be released in a future non-breaking patch version.

evanlinjin committed:
feat(chain)!: rm `get_chain_position` and associated methods

notmandatory commented:
On further discussion today at release planning call this PR was moved back into the 1.0 milestone if it can be completed and reviewed in time. If not it will have to wait for a 2.0 milestone because it required breaking changes to chain crate APIs that are exposed in the `Wallet` API.  

Alternatively we could remove the following `Wallet` functions and move any error types from chain into the core module which should insulate the wallet crate from this chain crate API changes. 

```rust
    /// Get a reference to the inner [`TxGraph`].
    pub fn tx_graph(&self) -> &TxGraph<ConfirmationBlockTime> {
        self.indexed_graph.graph()
    }

    /// Get a reference to the inner [`KeychainTxOutIndex`].
    pub fn spk_index(&self) -> &KeychainTxOutIndex<KeychainKind> {
        &self.indexed_graph.index
    }

    /// Get a reference to the inner [`LocalChain`].
    pub fn local_chain(&self) -> &LocalChain {
        &self.chain
    }
```

evanlinjin commented:
@notmandatory how does this solution compare with just giving wallet a major version bump?

notmandatory commented:
Removing the above functions and moving required chain error type to core should allow us to do breaking chain crate api changes without having to do a major wallet crate release. I also don't see why Wallet users need to access the inner chain types directly instead of using higher level functions like `transactions()`, or we can add new helper functions without breaking any APIs. 

But all that said if it's less risky to just keep everything as is and do a 2.0 release in 6 mo or so I'd be fine with that too.

evanlinjin committed:
feat(chain)!: Add `run_until_finished` methods

Add `run_until_finished` methods for `TxAncestors` and `TxDescendants`.
This is useful for traversing until the internal closure returns `None`.

Signatures of `TxAncestors` and `TxDescendants` are changed to enforce
generic bounds in the type definition.

evanlinjin committed:
test(chain): Add transitive anchor tests

evanlinjin commented on crates/wallet/src/test_utils.rs:
```diff
@@ -224,37 +224,63 @@ pub fn feerate_unchecked(sat_vb: f64) -> FeeRate {
+    //         transitively: None,
+    //     }
+    // };
+    // receive_output(wallet, value, anchor)
```
woops.

evanlinjin commented:
All that remains is docs and updating the PR description.

evanlinjin committed:
docs: Add ADR for `O(n)` canonicalization algorithm

evanlinjin committed:
perf(chain): add benchmarks for canonicalization logic

This is mostly taken from #1735 except we inline many of the functions
and test `list_canonical_txs`, `filter_chain_unspents` and
`filter_chain_txouts` on all scenarios.

CI and README is updated to pin `csv`.

Co-authored-by: valued mammal <valuedmammal@protonmail.com>

evanlinjin committed:
feat(chain): Derive `Clone` on `IndexedTxGraph`

evanlinjin commented on README.md:
```diff
@@ -78,6 +78,10 @@ cargo update -p tokio --precise "1.38.1"
+
+# Additional pins for running tests
+cargo update -p csv --precise "1.3.0"
+cargo update -p unicode-width --precise "0.1.13"
```
This is unfortunate. I think we should do #1398.

ValuedMammal commented:
Concept ACK, I'm still reviewing

Anecdotally I can attest that computation scales linearly with `TxGraph` size.

LLFourn commented on README.md:
```diff
@@ -78,6 +78,10 @@ cargo update -p tokio --precise "1.38.1"
 cargo update -p indexmap --precise "2.5.0"
 cargo update -p security-framework-sys --precise "2.11.1"
+
+# Additional pins for running tests
```
aren't all these pins for running tests...? I mean if you are just building a library then you likely won't have many of these dependencies.

evanlinjin replied:
Good point.

evanlinjin replied:
Removed this comment in the latest push.

LLFourn commented on crates/chain/src/tx_graph.rs:
```diff
@@ -134,16 +140,19 @@ impl<A: Ord + Clone> From<TxUpdate<A>> for TxGraph<A> {
 
-    // This atrocity exists so that `TxGraph::outspends()` can return a reference.
-    // FIXME: This can be removed once `HashSet::new` is a const fn.
+    txs_by_anchor: BTreeSet<(u32, Txid)>,
```
:question: 
```suggestion
    highest_confirmation_heights: BTreeSet<(u32, Txid)>,
```

evanlinjin replied:
Renamed to `txs_by_highest_conf_height`.

LLFourn commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,319 @@
+
+    /// Whether this transaction is already visited.
+    fn is_visited(&self, txid: Txid) -> bool {
+        self.canonical.contains_key(&txid) || self.not_canonical.contains(&txid)
```
`is_visited`  is a bit of a lie. `scan_anchors` is visiting the tx but it doesn't always make it `is_visited`. This should be `is_canonicalized`.

evanlinjin replied:
Changed in the latest push.

LLFourn commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -591,10 +591,84 @@ fn test_tx_conflict_handling() {
+                confirmed: Amount::from_sat(8000),
+                ..Default::default()
+            }
+        }
```
There seems to be a test missing for an anchored tx but not confirmed tx being canonical (or not). The orphaned block tests we have all are about conflicts.

P.S. this file is test_tx_graph_conflicts.rs but maybe it should just be `test_tx_graph_canonicalization.rs` .

evanlinjin replied:
In terms of naming, I think conflicts and canonicalization are mostly synonymous. We don't need to canonicalize if we don't have conflicts.

evanlinjin replied:
Test added.

LLFourn reviewed (approved):
ACK d107e4be516b1282cd5301322abbb15257934a40

Really excellent work. Thanks so much for doing an in-depth ADR. Made minor comments and a request for a test.

jirijakes commented:
I realize it's a lot of code but would it be possible to turn commits https://github.com/bitcoindevkit/bdk/pull/1670/commits/759be4c8f55adb266559fbdec90e1dc5ff96f9e6 and https://github.com/bitcoindevkit/bdk/pull/1670/commits/3577c5f25f3725f706627cf1786a24860cce63aa into one? The former commit introduces some changes (e. g. creates `canonical_iter.rs`) but the latter one changes them, which makes it more difficult to review commit by commit.

thunderbiscuit commented:
Thanks for the fantastic ADR @evanlinjin!

Perfect for me to stay up to speed on the changes without requiring a full code review, which I would not have had time to do! 🔥🔥

jirijakes committed:
refactor(wallet): Reuse chain position instead of obtaining new one

In `Wallet::preselect_utxos()`, the code used to obtain chain position
of the UTXO's transaction from the graph, however the chain position
is already recorded within the UTXO's representation (`LocalOutput`).
This patch reuses the existing chain position instead of obtaining a
fresh one.

nymius commented on crates/chain/benches/canonicalization.rs:
Shouldn't this be separated in another PR to review (and merge) first in master? Then benchmark test could be run against master before this PR get merge and have a reproducible way to check the performance increment after it does.

evanlinjin replied:
I've cherry-picked the benchmark commits into a separate branch that is based on master here: https://github.com/evanlinjin/bdk/tree/fix/1665-master-bench.

The problem is that the benchmarks on master takes so long that I don't think there is a point to check the performance increment against master.

ValuedMammal commented on crates/chain/src/lib.rs:
```diff
@@ -43,6 +43,8 @@ pub mod tx_graph;
 mod chain_oracle;
 pub use chain_oracle::*;
+mod canonical_iter;
+pub use canonical_iter::*;
```
I would have opted for less public visibility for most of the items in here.

evanlinjin replied:
We have an iterator `CanonicalIter` which should be public since it's returned by `TxGraph::canonical_iter`. `<CanonicalIter as Iterator>::Item` contains `CanonicalReason` so that needs to be public. Subsequently, `LastSeenIn` also needs to be public.

Otherwise, I've tried to reduce the publicly exposed items in the latest push.

ValuedMammal commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,319 @@
+/// Represents when and where a given transaction is last seen.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, core::hash::Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum LastSeenIn {
```
Name suggestion to avoid overloading the term last-seen. Also why does it implement serde traits?

```suggestion
pub enum ObservedIn {
```

evanlinjin replied:
Changed.

ValuedMammal commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,319 @@
+            LastSeenIn::Mempool(last_seen) => ChainPosition::Unconfirmed {
+                last_seen: Some(last_seen),
+            },
+            LastSeenIn::Block(_) => ChainPosition::Unconfirmed { last_seen: None },
```
I would have expected this to be not canon. based on my reading of the definition of `CanonicalTx`.

evanlinjin replied:
The docs of `CanonicalTx` needs to be changed. If a tx is not seen in the mempool, but we've seen it in a conflicting block, we know that the tx is seen by the network so we should still consider it canonical.

evanlinjin replied:
Updated in the latest push.

ValuedMammal commented on docs/adr/0003_canonicalization_algorithm.md:
```diff
@@ -0,0 +1,83 @@
+
+## Context and Problem Statement
+
+The Wizardsardine BDK audit uncovered the severity of the performance issues in the original canonicalization logic. The problem is especially severe for wallet histories with many unconfirmed and conflicting transactions. This can be a dDos vector if BDK is used in server-side applications. The time complexity of the original canonicalization logic is `O(n^2)`.
```
Consider using math notation, e.g. $O(n^2)$

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions

evanlinjin replied:
I didn't know this was a thing. Thanks!

ValuedMammal reviewed (approved):
ACK d107e4be516b1282cd5301322abbb15257934a40

📌 Comments are mostly for discussion and no action is required unless you plan to retouch.

notmandatory commented on docs/adr/0003_canonicalization_algorithm.md:
```diff
@@ -0,0 +1,83 @@
+
+## Context and Problem Statement
+
+The Wizardsardine BDK audit uncovered the severity of the performance issues in the original canonicalization logic. The problem is especially severe for wallet histories with many unconfirmed and conflicting transactions. This can be a dDos vector if BDK is used in server-side applications. The time complexity of the original canonicalization logic is `O(n^2)`.
```
⛏️ Should include a link to the audit report.
```suggestion
The [2024 Wizardsardine BDK code audit](https://gist.github.com/darosior/4aeb9512d7f1ac7666abc317d6f9453b) uncovered the severity of the performance issues in the original canonicalization logic. The problem is especially severe for wallet histories with many unconfirmed and conflicting transactions. This can be a dDos vector if BDK is used in server-side applications. The time complexity of the original canonicalization logic is `O(n^2)`.
```

nymius commented on crates/chain/benches/canonicalization.rs:
```diff
@@ -0,0 +1,263 @@
+    });
+}
+
+pub fn nested_conflicts(c: &mut Criterion) {
```
I'm having problems grokking the idea of "nested" conflicts. The code below creates the following transaction dependency tree, right?
```text
	    A                  B
	   /|\                /|\ 
	  / | \              / | \
	 /  |  \            /  |  \
	/   |   \          /   |   \
       /    |    \        /    |    \
      /     |     \      /     |     \
     /      |      \    /      |      \
    tx1...tx2099   tx1050     tx2...tx2100
```
Then the "answers" to the assert would be:
`nested_conflicts_unconfirmed::list_canonical_txs` -->  `A,B,tx1,tx2`
`nested_conflicts_unconfirmed::filter_chain_txouts` --> `TxOut(A), TxOut(B)`
`nested_conflicts_unconfirmed::filter_chain_unspents` --> `0 as TxOut(A) is spent by (unconfirmed) tx1 and TxOut(B) is spent by (unconfirmed) tx2` 

So, why "nested"?

evanlinjin replied:
Good catch, this benchmark is implemented wrong.

evanlinjin replied:
Fixed the test in the latest push.

notmandatory commented on docs/adr/0003_canonicalization_algorithm.md:
```diff
@@ -0,0 +1,83 @@
+* Transitive anchors are handled correctly.
+* We can efficiently mark large subsets as canonical/non-canonical.
+
+**Cons:** ?
```
```suggestion
**Cons:** More API and code changes.
```

evanlinjin replied:
I don't think the API surface changes are substantial - it's still mostly the same. Definitely more code changes though. However, I don't think this is a con in terms of architecture? Just the implementer and reviewers need to do more work.

evanlinjin committed:
refactor(chain)!: Rename `LastSeenIn` to `ObservedIn`

Also removed extra derives on `ObservedIn` and updated docs for
`CanonicalTx`.

nymius reviewed (commented):
cACK

These are my results for running the benchmark on this PR:
```text
many_conflicting_unconfirmed::list_canonical_txs [529.28 us 533.25 us 537.34 us]
many_conflicting_unconfirmed::filter_chain_txouts [560.02 us 566.69 us 574.36 us]
many_conflicting_unconfirmed::filter_chain_unspents [570.67 us 576.12 us 581.60 us]
many_chained_unconfirmed::list_canonical_txs [1.8421 ms 1.8820 ms 1.9277 ms]
many_chained_unconfirmed::filter_chain_txouts [2.5919 ms 2.6256 ms 2.6601 ms]
many_chained_unconfirmed::filter_chain_unspents [2.7202 ms 2.7679 ms 2.8201 ms]
nested_conflicts_unconfirmed::list_canonical_txs [505.92 us 509.53 us 513.44 us]
nested_conflicts_unconfirmed::filter_chain_txouts [502.93 us 511.17 us 519.27 us]
nested_conflicts_unconfirmed::filter_chain_unspents [485.62 us 488.70 us 492.61 us]
```

### Nits:
1. `crates/chain/tests/test_indexed_tx_graph.rs` lines 529, 536 and 566 still exhibit the old `get_chain_position` name.
2. The `CHANGELOG NOTICE` mentions it removes the `try_get_chain_position` twice. Should it be `try_get_chain_position` first and then `try_get_chain_spend` second?
3. I stumbled with the following _warning_ while running the bench code:
```bash
warning: the following packages contain code that will be rejected by a future version of Rust: criterion v0.2.11
note: to see what the problems were, use the option `--future-incompat-report`, or run `cargo report future-incompatibilities --id 1`
```

evanlinjin committed:
test(chain): Add canonicalization test

Tx anchored in orphaned block and not seen in the mempool should be
canon.

evanlinjin commented:
```
warning: the following packages contain code that will be rejected by a future version of Rust: criterion v0.2.11
note: to see what the problems were, use the option `--future-incompat-report`, or run `cargo report future-incompatibilities --id 1`
```

@nymius Yes, this warning is annoying. Right now our MSRV policy is enforced in the tests (which I don't think is a good idea) so we are forced to use a very old version of `criterion`. I think we should do bitcoindevkit/bdk_wallet#101 and update our dev dependencies.

jirijakes commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2014,20 +2026,22 @@ impl Wallet {
+                    Some(CanonicalTx {
+                        chain_position,
+                        tx_node,
+                    }) => (chain_position, tx_node.tx.clone()),
```
In 5456f8b4 feat(chain)!: rm `get_chain_position` and associated methods

```suggestion
                    }) => (chain_position, &tx_node.tx),
```

evanlinjin replied:
This line no longer exists.

jirijakes commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -2014,20 +2026,22 @@ impl Wallet {
+            .list_canonical_txs(&self.chain, chain_tip)
+            .map(|canon_tx| (canon_tx.tx_node.txid, canon_tx))
+            .collect::<HashMap<Txid, _>>();
+
```
In https://github.com/bitcoindevkit/bdk/commit/5456f8b4cd3fabaf1215d2bdec06bd0eae67a86f feat(chain)!: rm get_chain_position and associated methods:

Is it necessary to build list of canonical transactions here?

If it's just to have `chain_position` available, then that should not be necessary, chain position is already contained in `u.0` (`LocalOutput`) and can be extracted from there (see #1758).

If I remove `canon_txs` and apply the same change as in #1758, tests still pass.

```rs
        //    must_spend <- manually selected utxos
        //    may_spend  <- all other available utxos
        let mut may_spend = self.get_available_utxos();

        may_spend.retain(|may_spend| {
            !manually_selected
                .iter()
                .any(|manually_selected| manually_selected.utxo.outpoint() == may_spend.0.outpoint)
        });
        let mut must_spend = manually_selected;

        // NOTE: we are intentionally ignoring `unspendable` here. i.e manual
        // selection overrides unspendable.
        if *manually_selected_only {
            return (must_spend, vec![]);
        }

        let satisfies_confirmed = may_spend
            .iter()
            .map(|u| -> bool {
                let txid = u.0.outpoint.txid;
                let tx = match self.indexed_graph.graph().get_tx(txid) {
                    Some(tx) => tx,
                    None => return false,
                };

                // Whether the UTXO is mature and, if needed, confirmed
                let mut spendable = true;
                let chain_position = u.0.chain_position;

```

evanlinjin replied:
Good catch, included your commit.

jirijakes commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -1833,9 +1832,28 @@ impl Wallet {
+            .graph()
+            .list_canonical_txs(&self.chain, chain_tip)
+            .filter(|canon_tx| prev_txids.contains(&canon_tx.tx_node.txid))
+            .take(prev_txids.len())
```
In 5456f8b4 feat(chain)!: rm `get_chain_position` and associated methods:

`.take(prev_txids.len())` should not be needed. After filtering there can't be more than `prev_txids.len()` transactions. And then `filter_map` could be used.

If I make both `prev_txids` and `confirmation_heights` empty, no test in `wallet` fails. It seems like some execution paths are not covered by tests, so it is hard to tell whether logic of the implementation changed or not.

evanlinjin replied:
It's for a small performance gain. We may not return anything since `filter` filters it out. However, the `CanonicalIter` will continue to try canonicalize. Will add a comment.

> If I make both prev_txids and confirmation_heights empty, no test in wallet fails. It seems like some execution paths are not covered by tests, so it is hard to tell whether logic of the implementation changed or not.

Lets create a ticket since this is out of scope for this PR. However, currently BDK does not handle median-time-past properly.

evanlinjin replied:
I decided to include a comment on #642 instead.

jirijakes commented on crates/chain/tests/test_indexed_tx_graph.rs:
In this file, there are still some references to `get_chain_position` which does not exist anymore.

oleonardolima replied:
Yes, some are still at L529 (the test name), L536, L566.

evanlinjin replied:
Should be somewhat resolved by latest push.

jirijakes reviewed (commented):
Thanks for rerebasing and resquashing :pray:

A few comments from first round of reviewing. I have not yet reviewed the algorithm itself and ADR, I intend to do so during Saturday.

nymius commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+    chain: &'g C,
+    chain_tip: BlockId,
+
+    pending_anchored: Box<dyn Iterator<Item = (Txid, Arc<Transaction>, &'g BTreeSet<A>)> + 'g>,
```
`pending` in bitcoin context may be ambiguous. From what I understand, `pending` here refers to those transactions that haven't been canonicalized yet. Could the struct be documented considering this?

jirijakes replied:
:+1: 

Or simply `anchored_iter` and `last_seen_iter`.

evanlinjin replied:
How about `unprocessed_anchored_txs`?

nymius replied:
`unprocessed` seems right

nymius reviewed (commented):
Great work!

Final NIT. From PR branch:
```bash
rg -l "canoncical|canoncial|canonial" | xargs sed -i "s/\(canoncical\|canoncial\|canonial\)/canonical/g"
```

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+    }
+}
+
+impl<'g, A: Anchor, C: ChainOracle> Iterator for CanonicalIter<'g, A, C> {
```
```suggestion
impl<A: Anchor, C: ChainOracle> Iterator for CanonicalIter<'_, A, C> {
```

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+                .chain
+                .is_block_in_chain(anchor.anchor_block(), self.chain_tip)?;
+            if in_chain_opt == Some(true) {
+                self.mark_canonical(tx, CanonicalReason::from_anchor(anchor.clone()));
```
:wrench: I believe cloning should not be necessary here. However, without further refactoring it might be needed on a higher level. It could be tried in a follow-up.

evanlinjin replied:
Are you thinking about doing this with references or by making `Anchor` be bound by `Copy`?

jirijakes replied:
I meant reference. But of course, unnecessary in this PR. I may give it a try after it's merged.

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+    /// Marks a transaction and it's ancestors as canoncial. Mark all conflicts of these as
+    /// `not_canonical`.
+    fn mark_canonical(&mut self, tx: Arc<Transaction>, reason: CanonicalReason<A>) {
+        let starting_txid = tx.compute_txid();
```
Txid is already available everytime `mark_canonical` is called. We could avoid recomputing it again.

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+                Some(())
+            },
+        )
+        .for_each(|_| {})
```
```suggestion
        .run_until_finished()
```

oleonardolima replied:
Yes, +1 on this.

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+
+    /// Marks a transaction and it's ancestors as canoncial. Mark all conflicts of these as
+    /// `not_canonical`.
+    fn mark_canonical(&mut self, tx: Arc<Transaction>, reason: CanonicalReason<A>) {
```
If this method is called `canonicalize`, would it make sense? Is the meaning of `is_canonicalized` the same as “gone through `mark_canonical`”?

I am not necessarily suggesting to rename it, I am trying to understand.

evanlinjin replied:
This whole algorithm is a canonicalization algorithm. So I'm that sense, we are canonicalizing everything and the name `canonicalize` will be ambiguous. 

We are explicitly meeting a tx as canonical here. We make use of the transitive nature of a canonical tx to mark more things. 

`is_canonicalized` returns whether we have an answer for a given tx. If so, we can skip it.

jirijakes commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -0,0 +1,251 @@
+            anchors
+                .iter()
+                .last()
+                .unwrap()
```
Maybe add `debug_assert` to check that `anchors` is not empty?

evanlinjin replied:
What are we trying to achieve with this? Would changing `.unwrap()` to `.expect()` suffice?

evanlinjin replied:
Oh, at the top of the method?

evanlinjin replied:
I guess the problem will still get picked up by the `.expect()`?

jirijakes replied:
Yes, at the top. True, in this specific case, the expect would work similarly and adding `debug_assert` now is not worth it.

(I came to believe that in general, putting `debug_assert` around the code is a [good practice](https://tratt.net/laurie/blog/2023/rusts_two_kinds_of_assert_make_for_better_code.html) for documentation purposes and for increasing a chance of discovering bugs.)

evanlinjin replied:
That's a great read and thanks for sharing.

jirijakes commented on crates/chain/src/tx_graph.rs:
```diff
@@ -624,14 +634,36 @@ impl<A: Clone + Ord> TxGraph<A> {
+                }
+                old_top_h = top_anchor_height(e.get());
+                let is_changed = e.get_mut().insert(anchor.clone());
+                new_top_h = top_anchor_height(e.get()).expect("must have anchor");
```
🔧 It is a bit hard to understand that what this does is to determine whether a newly inserted anchor has higher height than any of previously existing anchors. It is especially confusing that later the two values are compared as `old_top_h != Some(new_top_h)`.

A comment may help. Or perhaps to find `new_top_h`, compare `old_top_h` with `anchor` directly instead of finding BTreeSet's max value again: either just-inserted `anchor` will be higher or `old_top_h`.

evanlinjin replied:
Good catch. I tried reorganizing this logic in 87528963805da281c2f674eba63e22a075613117. Let me know what you think.

jirijakes reviewed (commented):
Reviewed the new algorithm and ADR. Left a few comments but nothing extremely important.

Great work!

notmandatory reviewed (approved):
ACK 664bb9ae97a5b5e347c5b2bb9c989f0bfb5fa8d3

Great job with the ADR, it's a big help for understand the logic. The benchmarks are also a first for the project and do a fantastic job quantifying the value of this change.

oleonardolima commented on crates/chain/src/tx_graph.rs:
```diff
@@ -138,12 +142,15 @@ impl<A: Ord + Clone> From<TxUpdate<A>> for TxGraph<A> {
     anchors: HashMap<Txid, BTreeSet<A>>,
     last_seen: HashMap<Txid, u64>,
 
+    txs_by_highest_conf_heights: BTreeSet<(u32, Txid)>,
+    txs_by_last_seen: BTreeSet<(u64, Txid)>,
```
nit: we could probably use `txids_by_...` instead, wdyt ?

evanlinjin replied:
I don't think this will make a significant difference to the readability of the codebase.

notmandatory commented:
Might need a rebase.

oleonardolima commented on crates/chain/tests/test_tx_graph_conflicts.rs:
```diff
@@ -595,6 +597,7 @@ fn test_tx_conflict_handling() {
 
     for scenario in scenarios {
         let (tx_graph, spk_index, exp_tx_ids) = init_graph(scenario.tx_templates.iter());
+        let exp_tx_ids = dbg!(exp_tx_ids);
```
nit: Did you mean to keep this `dbg!()` ?

oleonardolima commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -214,36 +213,39 @@ impl<A: Clone> CanonicalReason<A> {
     }
 
     /// Constructs a [`CanonicalReason`] from a `last_seen` value.
```
nit: I guess the docs need an update too.

``` suggestion
    /// Constructs a [`CanonicalReason`] from a `observed_in` value.
```

oleonardolima commented on crates/chain/src/canonical_iter.rs:
```diff
@@ -214,36 +213,39 @@ impl<A: Clone> CanonicalReason<A> {
+                observed_in: last_seen,
+                ..
+            } => Self::ObservedIn {
+                observed_in: *last_seen,
```
Is it intentional to keep using `last_seen` here instead of all `observed_in`?

evanlinjin replied:
Fixed

oleonardolima reviewed (commented):
Concept ACK

Great work! I don't have much experience with the `bdk_chain`, so I'll need another round to wrap my mind around the changes, but the docs and ADR already helped a lot 😁.

I left a few comments with some possibly small applicable nits.

evanlinjin committed:
chore(chain): Address `CanonicalIter` nitpicks

evanlinjin committed:
refactor(chain): Reorganize `TxGraph::insert_anchor` logic for clarity

evanlinjin committed:
docs(wallet): Explain `.take` usage

evanlinjin committed:
chore: Fix typos

evanlinjin committed:
test(chain): Update test docs to stop referencing `get_chain_position`

ValuedMammal reviewed (approved):
ACK 956d0a93bab36fd919e81718832fbc1033b677e9

nymius reviewed (approved):
ACK 956d0a93

oleonardolima reviewed (commented):
utACK 956d0a93bab36fd919e81718832fbc1033b677e9

jirijakes reviewed (approved):
ACK 956d0a93bab36fd919e81718832fbc1033b677e9

:fire:

notmandatory commented:
Thanks @evanlinjin and @ValuedMammal for all the work on getting this coded, documented, and benchmark tested, and to everyone who helped with review!
